### DISCUSSION TITLE
A good way to clean all the test environment?

### DISCUSSION THREAD

**Discussion by danielfoguelman** [pytest-dev/pytest#10353] [Q&A]
Hi! During a pull request on a company feature I found that there is a hidden dependency between my test cases. I wonder if there is a clean and happy way to clean the python environment between test file executions.
I would like to unload all modules / restore env variables / etc.
Any good ideas?

**Comment by portyu9** [comment]
Hi @danielfoguelman 

pytest deliberately runs tests in the same Python interpreter by default, so there is no reliable "reset the entire Python environment between test files" option.

In particular, trying to remove everything from `sys.modules` is not a complete reset: existing references, imported objects, C extensions, global caches, threads, etc. can all survive.

The usual fix is to isolate the state that each test changes:

- use fixtures with teardown/`yield` for resources
- use `monkeypatch.setenv()` / `monkeypatch.delenv()` for environment variables
- use `monkeypatch.setattr()` for global/module attributes
- clear application caches explicitly
- use `tmp_path` for filesystem state

For example:

```python
def test_something(monkeypatch):
    monkeypatch.setenv("MY_SETTING", "test")
```

pytest restores the environment change automatically after the test.

If you genuinely need a **completely fresh interpreter**, process isolation is the correct boundary: run the relevant test file/group in a separate pytest subprocess.

I would also treat the ordering dependency as something worth fixing rather than relying on a global cleanup mechanism; tests should ideally pass regardless of which test ran before them.

Please mark this answer as accepted if it helped, thank you.
